core/rpc/src/state/state_full.rs

```diff
@@ -83,58 +84,46 @@ impl<B, E, Block: BlockT, RA> FullState<B, E, Block, RA>
 		to: Option<Block::Hash>
 	) -> Result<QueryStorageRange<Block>> {
 		let to = self.block_or_best(to).map_err(client_err)?;
-		let from_hdr = self.client.header(&BlockId::hash(from)).map_err(client_err)?;
-		let to_hdr = self.client.header(&BlockId::hash(to)).map_err(client_err)?;
-		match (from_hdr, to_hdr) {
-			(Some(ref from), Some(ref to)) if from.number() <= to.number() => {
-				// check if we can get from `to` to `from` by going through parent_hashes.
-				let from_number = *from.number();
-				let blocks = {
-					let mut blocks = vec![to.hash()];
-					let mut last = to.clone();
-					while *last.number() > from_number {
-						let hdr = self.client
-							.header(&BlockId::hash(*last.parent_hash()))
-							.map_err(client_err)?;
-						if let Some(hdr) = hdr {
-							blocks.push(hdr.hash());
-							last = hdr;
-						} else {
-							return Err(invalid_block_range(
-								Some(from),
-								Some(to),
-								format!("Parent of {} ({}) not found", last.number(), last.hash()),
-							))
-						}
-					}
-					if last.hash() != from.hash() {
-						return Err(invalid_block_range(
-							Some(from),
-							Some(to),
-							format!("Expected to reach `from`, got {} ({})", last.number(), last.hash()),
-						))
-					}
-					blocks.reverse();
-					blocks
-				};
-				// check if we can filter blocks-with-changes from some (sub)range using changes tries
-				let changes_trie_range = self.client
-					.max_key_changes_range(from_number, BlockId::Hash(to.hash()))
-					.map_err(client_err)?;
-				let filtered_range_begin = changes_trie_range
-					.map(|(begin, _)| (begin - from_number).saturated_into::<usize>());
-				let (unfiltered_range, filtered_range) = split_range(blocks.len(), filtered_range_begin);
-				Ok(QueryStorageRange {
-					hashes: blocks,
-					first_number: from_number,
-					unfiltered_range,
-					filtered_range,
-				})
-			},
-			(from, to) => Err(
-				invalid_block_range(from.as_ref(), to.as_ref(), "Invalid range or unknown block".into())
-			),
+		let from = self.client.header_metadata(from).map_err(client_err)?;
```

Comment:
Previously the error message in case a block is missing was much more friendly than this one. I guess right now you will get quite cryptic `ClientError`, where previously you were actually getting errors in consistent format. I really think we should get this behaviour back and perhaps even add a test for this case.

Reply:
This is exactly the same as the original implementation, printing something like `header not found in db: {hash}`, just compare the lines. So, `much more friendly and consistent` sounds like an exaggeration  :-).

But I guess we can use `invalid_block_range` everywhere and print both hashes/number.